File: binary_tree/1_binary_tree.cpp
/* ************************************************************************
> File Name:     1_binary_tree.cpp
> Author:        xmw01
> Created Time:  Sun 14 May 2023 05:46:33 PM CST
> Description:   
 ************************************************************************/


struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(NULL), right(NULL) {}
}


// recursive traversal
// 递归遍历

// 前序遍历
class Solution
{
public:
    void traversal(TreeNode* cur, vector<int>& vec){
        if(cur == NULL) return;
        vec.push_back(cur->val);   //中
        traversal(cur->left, vec); //左
        traversal(cur->right, vec);//右
    }
    vector<int> perorderTraversal(TreeNode* root){
        vector<int> result;
        traversal(root, result);
        return result;
    }
};

// 中序遍历
class Solution
{
public:
    void traversal(TreeNode* cur, vector<int>& vec){
        if(cur == NULL) return;
        traversal(cur->left, vec);  //左
        vec.push_back(cur->val);    //中
        traversal(cur->right, vec); //右
    }
    vector<int> midorderTraversal(TreeNode* root){
        vector<int> result;
        traversal(root, result);
        return result;
    }
};

// 后序遍历
class Solution
{
public:
    void traversal(TreeNode* cur, vector<int>& vec){
        if(cur == NULL) return;
        traversal(cur->left, vec);  //左
        traversal(cur->right, vec); //右
        vec.push_back(cur->val);    //中
    }
    vector<int> postorderTraversal(TreeNode* root){
        vector<int> result;
        traversal(root, result);
        return result;
    }
};

File: binary_tree/python/1_binaryTree.py
class TreeNode(object):
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

def constructBinaryTree(nums):
    Treelist = [None for _ in range(len(nums))]

    for i in range(len(nums)):
        if nums[i] != None:
            node = TreeNode(nums[i])
            Treelist[i] = node
        if i == 0: root = node

    i = 0
    #注意结束的规则如下
    while i * 2 + 2 < len(nums):
        if Treelist[i] != None:
            # 和c++ 构建一样，这个是关键逻辑
            Treelist[i].left = Treelist[2 * i + 1]
            Treelist[i].right = Treelist[2 * i + 2]
            i += 1
    #返回根节点
    return root

if __name__ == '__main__':

    nums = [4,1,6,0,2,5,7,None,None,None,3,None,None,None,8]
    root = constructBinaryTree(nums)
    print(root)
